### ISSUE TITLE
Getting exception "Expected 1 row affected, actual 4." when deleting RecordSet with multiple Records

### ISSUE THREAD

**Issue by kevinsmith** [atlasphp/Atlas.Table#9]
I'm not sure if this belongs in the ORM project, the Mapper project, or this one, so I'm filing it in the project where the exception was thrown.

I've got a simple table setup, no relationships. Running a select query to find all records with a given user ID and email address, then deleting them all. Code looks like this:

```php
        $recordSet = $this->atlas
            ->select(TableName::class)
            ->where('user_id =', $userId)
            ->andWhere('email_address =', $emailAddress)
            ->fetchRecordSet();

        $recordSet->setDelete();

        $atlas->persistRecordSet($recordSet);
```

When I run it, it deletes all the matching records as expected, but it throws an exception coming from https://github.com/atlasphp/Atlas.Table/blob/faeec8db3e2dacb432144563cd405b17600f4e6e/src/Table.php#L252

that reads "Expected 1 row affected, actual 4." in an example that has 4 matching rows.

Not sure what's going on here, but it looks like `$pdoStatement` is grabbing the count for all affected rows and for some reason the caller expects there to only be 1 affected row.

**Comment by froschdesign** [comment]
The related method is:

https://github.com/atlasphp/Atlas.Table/blob/faeec8db3e2dacb432144563cd405b17600f4e6e/src/Table.php#L242

But it seems that all rows of the recordset will be deleted at once:

> PDOStatement::rowCount() returns the number of rows affected by the last DELETE, INSERT, or UPDATE statement executed by the corresponding PDOStatement object.

https://www.php.net/manual/pdostatement.rowcount.php

**Comment by kevinsmith** [comment]
Aw, dammit. It was because I didn't have a primary key on the table, so Atlas was effectively selecting all the rows for deletion when it went to delete the first record. So the "expected" number here was actually correct, it should've just been deleting one row at a time, but Atlas didn't know which row it was deleting so (I believe) it was trying to delete them all. Oof. That seems like a major data integrity issue.

@pmjones Does Atlas require a primary key?

**Comment by pmjones** [comment]
Hey @kevinsmith !

> Does Atlas require a primary key?

Atlas "as a project" does not; e.g., Atlas.Pdo and Atlas.Query will not raise that kind of exception.

But some the Row-related methods *do* expect to only ever affect a single row; I mean, that's the whole point of "insert a row", "delete a row", etc.  Thus the exception when more than one row is affected; it's definitely outside the expectation of working with a single row.

Looking back over this, though, I'm surprised that it happened here. I thought it *did* require a primary key, but apparently it does not. Cf. Table::deleteRowPrepare() --

```php
        $delete = $this->delete();
        foreach (static::PRIMARY_KEY as $primaryCol) {
            $delete->where("{$primaryCol} = ", $row->$primaryCol);
        }
```

There's no check to see if *is* a primary key at all. Same thing in updateRowPrepare().

That's a problem for sure, and I'll address it today.

Sorry for the problem, and thanks for raising the issue!

**Comment by pmjones** [comment]
@kevinsmith As a solution, I am contemplating adding a `hasWhere()` method to Atlas.Query; then in the delete/update prepare methods, throw an exception if no WHERE clause is present. That will allow both for primary keys, and for situations where there is no primary key but a table event has added some WHERE conditions. Does that sound sufficient for your purposes?

EDIT: Alternatively, I could make it so a primary key is absolutely required for the `*Row()*` methods. Thoughts?

**Comment by kevinsmith** [comment]
No worries! Thanks for taking a look at this.

This is a tough one, and I keep going back and forth on it. Conceptually, all that's necessary here is that the system be able to target a particular row when a particular row is intended. A primary key isn't the only way to accomplish that.

But given that's there's very little downside to having a primary key column in the table, it's the most performant way of targeting a particular row, and that in the vast majority of cases you would already need to have one, I'd say the cleanest and most straightforward approach here is to just require a primary key for these operations and throw an exception if one isn't present.

**Comment by pmjones** [comment]
> I'd say the cleanest and most straightforward approach here is to just require a primary key for these operations and throw an exception if one isn't present.

I tentatively agree. Let me see what I can work out, and thanks for your patience!

**Comment by pmjones** [comment]
@kevinsmith Take a look at the `primary-perform` branch, PR at #10 , and let me know if that soothes your woes.

@froschdesign Let me know if you have comments or concerns here!

**Comment by kevinsmith** [comment]
Thanks! Just took a look and found a small issue. Reported in that PR.

**Comment by pmjones** [comment]
Fixed by #10

**Comment by pmjones** [comment]
@kevinsmith I just released 1.4.0 with these changes in place. Let me know what else you might need!
